File: src/test/java/com/josephbaca/world/CoordinateTest.java
package com.josephbaca.world;

import static org.junit.jupiter.api.Assertions.*;
import static org.quicktheories.QuickTheory.qt;
import static org.quicktheories.generators.SourceDSL.integers;

import org.junit.jupiter.api.Test;

class CoordinateTest {

  @Test
  void testEquals() {
    assertEquals(Coordinate.of(0, 0),
                 Coordinate.of(0, 0));

    assertEquals(Coordinate.of(3, 5),
                 Coordinate.of(3, 5));

    assertNotEquals(Coordinate.of(3, 5),
                    Coordinate.of(5, 3));

    qt().forAll(
        integers().allPositive(),
        integers().allPositive())
        .check(
            (i, j) -> Coordinate.of(i, j).equals(Coordinate.of(i, j)));

    qt().forAll(
        integers().allPositive(),
        integers().allPositive())
        .assuming((i, j) -> !i.equals(j))
        .check(
            (i, j) -> !Coordinate.of(i, j).equals(Coordinate.of(j, i)));
  }

}

File: src/test/java/com/josephbaca/world/CoordinateTest.kt
package com.josephbaca.world

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Test
import org.quicktheories.QuickTheory.qt
import org.quicktheories.generators.SourceDSL.integers

internal class CoordinateTest {

    @Test
    fun testEquals() {
        assertEquals(
            Coordinate(0, 0),
            Coordinate(0, 0)
        )

        assertEquals(
            Coordinate(3, 5),
            Coordinate(3, 5)
        )

        assertNotEquals(
            Coordinate(3, 5),
            Coordinate(5, 3)
        )

        qt().forAll(
            integers().allPositive(),
            integers().allPositive()
        )
            .check { i, j -> Coordinate(i!!, j!!) == Coordinate(i, j) }

        qt().forAll(
            integers().allPositive(),
            integers().allPositive()
        )
            .assuming { i, j -> i != j }
            .check { i, j -> Coordinate(i!!, j!!) != Coordinate(j, i) }
    }

}